### ISSUE TITLE
[Bug Skip] test_sparse - test_index_select_exhaustive_index_large_xpu_float64 hangs on BMG (>600s timeout)

### ISSUE THREAD

**Issue by daisyden** [intel/torch-xpu-ops#4184]
## Bug Description

`test_index_select_exhaustive_index_large_xpu_float64` hangs indefinitely on Intel GPU (BMG / Battlemage). The test does not complete within a 600-second timeout (exit code 124), whereas the sibling test `test_index_select_empty_and_non_contiguous_index_xpu_float64` completes in under 120 seconds. This indicates the exhaustive-index path for large sparse tensors does not progress on XPU.

## Affected Tests

Cases:
op_ut,third_party.torch-xpu-ops.test.xpu.test_sparse_xpu,TestSparseXPU.test_index_select_exhaustive_index_large_xpu_float64

## Error Message

```
Test exceeded 600s hard timeout (exit code 124). No traceback produced - the process is still running (hung) when killed.
collected 1 item
test_sparse_xpu.py::TestSparseXPU::test_index_select_exhaustive_index_large_xpu_float64 <hangs>
=== EXIT: 124 ===
```

## Traceback

```
pytest_command:
cd /home/sdp/daisyden/pytorch/third_party/torch-xpu-ops/test/xpu && timeout 600 pytest -sv -p no:xdist "test_sparse_xpu.py::TestSparseXPU::test_index_select_exhaustive_index_large_xpu_float64"

Result: timed out / hung at both 120s and 600s limits (exit 124). No Python traceback - the test never returns.
```

## Root Cause Analysis

The exhaustive-index sparse `index_select` over a large tensor appears to hang on XPU. Verified in isolation (single test, `-p no:xdist`) at both a 120s and a 600s hard timeout; both produced exit code 124 (timeout), so this is a true hang rather than mere slowness, and not a parallel-execution artifact. The sibling test with empty / non-contiguous indices passes quickly (<120s), so the hang is specific to the large exhaustive-index code path. This may be an XPU sparse `index_select` kernel that fails to make progress (possible deadlock or extreme XPU->CPU fallback degeneracy) on large inputs.

## Versions

PyTorch: torch 2.13.0+xpu
Intel: intel-sycl-rt/intel-cmplr 2026.0.0, intel-pti 0.17.0
numpy: 1.23.2
GPU: Intel BMG (Battlemage), 23.91 GiB

**Comment by tszulist-hbn** [comment]
Here's the summary of the current investigation findings, The hang is **deterministic** — it occurs after ~5-19 iterations of sparse `index_select` followed by `.to_dense()` in a loop. The test's `assertEqual()` internally calls `.to_dense()` to compare sparse vs dense tensors, which triggers it.

**Minimal reproducer:**
```python
t_sparse = torch.randn((100, 50, 3, 3), device='xpu').to_sparse()
idx = torch.tensor([0, 1, 2], device='xpu')
for i in range(10):
    result = t_sparse.index_select(2, idx)
    dense = result.to_dense()  # Hangs after ~4-6 iterations
```

### What we ruled out

1. **Not the group_reduce workgroup alignment issue** — while initially PR #4228's seemed to to be the solution, eventually occurred it is correct but fixed a different problem (AR: report a separate issue).
2. **Not missing application-level synchronization** — Adding `torch.xpu.synchronize()`, `c10::xpu::syncStreamsOnDevice()`, and `getCurrentSYCLQueue().wait()` at all return paths in both `Indexing.cpp` and `SparseTensorMathKernels.cpp` did NOT fix the hang. It shifted the iteration count slightly (from ~18-19 to ~12) but the hang persists.

### Likely root cause

`IndexSelectSparse3Functor` has a pattern where multiple work-items execute parallel writes through pointer arithmetic in a for-loop:

```cpp
// Runs on multiple work-items in parallel
for (index_t i = 0; i < count; ++i) {
    *ptr_res_dim_indices_out++ = idx_idx;            // NON-ATOMIC WRITE
    *ptr_selected_dim_indices_out++ = *ptr_argsort_dim_indices_in++;
}
```

When work-items have overlapping memory regions, this creates a **intra-kernel memory race**. Queue-level synchronization (`wait()`) only ensures kernels complete sequentially — it doesn't prevent races *within* a kernel.

### Proposed next steps (in order of likelihood to succeed)

1. **Atomic operations** — Use `sycl::atomic_ref` for the writes in `IndexSelectSparse3Functor` to make them thread-safe
2. **Verify offset overlaps** — Add diagnostics to confirm whether work-item memory regions actually overlap (if they don't, the issue is elsewhere — possibly SYCL runtime resource exhaustion)
3. **Algorithm redesign** — Compare with the CUDA implementation and eliminate the dynamic-offset write pattern entirely

**Comment by tszulist-hbn** [comment]
### Root cause & fix

Root-caused on BMG. The hang was **two independent GPU-side deadlocks**, both surfacing with the same signature (host thread blocked on a D2H copy for a later `.item()` while a submitted kernel never completes). Both are fixed in #4228.

**1. `group_reduce` workgroup/subgroup misalignment** — reached from the sparse `index_select` reduction path. `group_reduce` assumes `wg_size % sg_size == 0`, but on BMG (where `max_sg_sz` is forced to 8) some launch configs produced non-aligned geometry, stalling the reduction. In release builds the guarding `SYCL_KERNEL_ASSERT` is compiled out, so it hangs instead of aborting.

**2. FP64 workgroup-scope local atomics** — reached from uncoalesced `.to_dense()` (`add_out_dense_sparse_kernel` → `index_add_` → `IndexFuncLargeIndexFunctor`). This XPU-only SLM write-coalescing optimization uses workgroup-scope 64-bit FP atomics (`atomicAddLocal<double>`), which are not reliably supported on BMG and deadlock the device.

Isolation pinned the second issue precisely:
- `double` + large nnz → hangs; `float` + large nnz → passes
- `double` + small nnz → passes
- `scatter_add_` (double, *global* atomics) → passes
- `.coalesce().to_dense()` → passes; direct uncoalesced `.to_dense()` → hangs

### Changes (in #4228)
- Align reduction / batch-kernel / scan workgroup sizes to a subgroup-size multiple before `group_reduce` (new `WorkGroupUtils.h` helper); drop the redundant assertion.
- In `IndexFuncLargeIndexFunctor`, skip the SLM local-atomic coalescing for `double`/`complex<double>` and accumulate via the global-memory atomic op instead (the same path CUDA uses). Other dtypes keep the SLM fast path.

### Result
`test_index_select_exhaustive_index_large_xpu_float64` now passes in ~0.7s
(previously >600s timeout).

**Comment by daisyden** [comment]
[agent_triage_result]

# Quick Label: intel/torch-xpu-ops#4184

https://github.com/intel/torch-xpu-ops/issues/4184

Root cause: Test issues ~54 exhaustive sparse `index_select` calls per run (test/test_sparse.py:1423-1447), each sorting indices and syncing via a device-reduction in the XPU sparse kernel (torch-xpu-ops/src/ATen/native/xpu/sycl/Indexing.cpp:2374-2418); the kernel structurally matches CUDA's bounded binary-search implementation (aten/src/ATen/native/cuda/Indexing.cu:1811-1988), so a true hang vs. a cumulative-cost timeout is unproven from static inspection.

| label | reason |
|---|---|
| `issue_type: Bug` | `github_type` field empty; title "[Bug Skip]" plus functionality-bug heuristic map to Bug. |
| `test_module: ut` | `extract.json` `test_module` is `ut`; reproduce_steps run pytest against `test_sparse_xpu.py`. |
| `module: sparse` | Targets sparse `index_select` dispatch (`SparseXPU` -> `index_select_sparse_xpu`, native_functions.yaml:9269-9281). |
| `P1` | Hang/timeout: process stayed alive but stuck past 600s (exit 124) twice, matches priority.md "Hang or timeout". |
| `dependency component: none` | Full call path resolves to torch-xpu-ops's own bounded kernel; no driver/IGC/Level_Zero/oneMKL/oneDNN frame. |
| `target_component: N/A` | Kernel's only loops are bounded binary search (Indexing.cpp:1979-2003) and a count-bounded copy (2291-2294); no infinite loop found. |
| `need_action: NEED_HUMAN` | `N/A` target_component -- hang vs. cumulative-cost timeout needs live reproduction, not resolvable statically. |
